File: ruby/lasagna/lasagna.rb
# frozen_string_literal: true

# Implementation of the Lasagna exercise in Ruby track on Exercism.
class Lasagna
  # expected time in the oven in minutes
  EXPECTED_MINUTES_IN_OVEN = 40

  # @param actual_minutes_in_oven [Integer]
  def remaining_minutes_in_oven(actual_minutes_in_oven)
    EXPECTED_MINUTES_IN_OVEN - actual_minutes_in_oven
  end

  # @param layers [Integer]
  # @return [Integer] preparation time in minutes
  def preparation_time_in_minutes(layers)
    layers * 2
  end

  # @param number_of_layers [Integer]
  # @param actual_minutes_in_oven [Integer]
  # @return [Integer] total time in minutes
  def total_time_in_minutes(number_of_layers:, actual_minutes_in_oven:)
    preparation_time_in_minutes(number_of_layers) + actual_minutes_in_oven
  end
end

File: elixir/lasagna/lib/lasagna.ex
defmodule Lasagna do
  @moduledoc """
  Implementation of the Lasagna exercise in Elixir track on Exercism.
  """
  @spec expected_minutes_in_oven :: Integer.t()
  def expected_minutes_in_oven do
    40
  end

  @spec remaining_minutes_in_oven(Integer.t()) :: Integer.t()
  def remaining_minutes_in_oven(elapsed) do
    expected_minutes_in_oven() - elapsed
  end

  @spec preparation_time_in_minutes(Integer.t()) :: Integer.t()
  def preparation_time_in_minutes(layers) do
    layers * 2
  end

  @spec total_time_in_minutes(Integer.t(), Integer.t()) :: Integer.t()
  def total_time_in_minutes(layers, minutes_in_oven) do
    preparation_time_in_minutes(layers) + minutes_in_oven
  end

  @spec alarm :: String.t()
  def alarm do
    "Ding!"
  end
end
